Instant Search: Add initial framework

gibrown opened the pull request:
#### Changes proposed in this Pull Request:
* Adds the build process and initial code for JP Search as you type front end Preact app
* Loads the front end assets when `JETPACK_SEARCH_PROTOTYPE` is defined as true
* The js code is a very minimal version that really doesn't do anything and will need to be heavily rewritten.

#### Testing instructions:
* Run `yarn build` and the distribution in `_inc/search/dist` should get created
* Add `define( "JETPACK_SEARCH_PROTOTYPE", true );` to your wp-config.php
* In order to load the code you need to have a JP Pro plan and have JP Search enabled.
* Load the front end of your test site and you should see `jp-search.bundle.js` get loaded. The front end site should not be affected otherwise.
* Try also `yarn build-production`. Front end js code should be pretty small.

#### Proposed changelog entry for your changes:
* No changelog for this.

gibrown committed:
First cut at putting some prototype jp search as you type code into Jetpack

gibrown commented:
Notes on things to figure out:
- i added the command `yarn build-search` but nothing works
 - I copied webpack.config to be one just for search, not sure if that is the right direction or not
 - i have no idea about what webpack.config.search.js should do
- hmmm, just merge in package.json to the jp one? no idea on any of this stuff
 - seem to be a bunch of version differences and we should try to get things in sync
- i copied the same dir structure into _inc/search. Probably the right location, not sure we want to keep this same structure though
- hook into translations somehow. merge into babel.config.js i assume
- there are a number of linter errors that I ignored

gibrown committed:
Merge more code over from the search-as-you-type repo

- took a guess at where things go for dependencies
- tried adding things to webpack build, but still not working

Commit:
Fix Webpack configuration for Jetpack Search

Commit:
Fix linting errors and rename files to JSX

jsnmoon commented:
Webpack config and linter violations have been fixed!

Commit:
Remove errant “i” package from package.json

gibrown commented:
Plan for enqueuing scripts is to rely on an is_defined for `JETPACK_SEARCH_PROTOTYPE`. I guess should probably go into a new class in the `modules/search` directory.

simison commented on webpack.config.search.js:
```diff
@@ -0,0 +1,23 @@
+	resolve: {
+		extensions: [ '.js', '.jsx' ],
+	},
+};
```
You might not need this extra Webpack config if you;
- add new entry points either to the main Webpack file (`webpack.config.js`) or the one geared towards Gutenberg extensions (`webpack.config.extensions.js`).
- Use Calypso-build directly from the command line (see [docs](https://github.com/Automattic/wp-calypso/tree/master/packages/calypso-build#usage) so that you won't need config.

Ideally there would be only one build command and one Webpack config. I still hope that sooner than later the two files would become one. :-)

jsnmoon replied:
I went ahead and merged the relevant parts of `webpack.config.search.js` into `webpack.config.js`.

simison replied:
See my related question about block editor extension: https://github.com/Automattic/jetpack/pull/12741#issuecomment-507424788

simison commented on package.json:
```diff
@@ -150,6 +152,13 @@
 		"react-dom": "16.8.6"
 	},
 	"devDependencies": {
+		"@automattic/calypso-build": "1.0.0",
```
Some of these devDependencies got prolly added here accidentally since they're in `dependencies` as well.

jsnmoon replied:
Fixed!

simison commented on package.json:
```diff
@@ -177,13 +186,16 @@
 		"nock": "10.0.6",
 		"node-wp-i18n": "1.2.3",
+		"preact": "^8.4.2",
+		"preact-portal": "^1.1.3",
```
Nit; note that dependencies in Jetpack are pinned (no `^`)

jsnmoon replied:
Fixed!

simison commented on package.json:
```diff
@@ -177,13 +186,16 @@
 		"nock": "10.0.6",
 		"node-wp-i18n": "1.2.3",
+		"preact": "^8.4.2",
+		"preact-portal": "^1.1.3",
```
Remember to also provide updated `yarn.lock`.

It should get updated when using `yarn add package-name` instead of `npm install`

jsnmoon replied:
`yarn.lock` has been generated!

simison commented on _inc/search/src/index.jsx:
```diff
@@ -0,0 +1,30 @@
+			initialValue={ initialValue }
+			grabFocus={ grabFocus }
+			SearchResults={ SearchResults }
+			api={ api }
```
Curious why `api` and `SearchResults` are passed down here instead of importing them directly in `SearchWidget`?

jsnmoon replied:
@notnownikki: Shall I go ahead and move the `api` and `SearchResults` imports into the `SearchWidget` component?

simison commented:
> need to define how this will merge onto wp.com and where the files will live there and get built

I'm lacking background/context but suppose that the editor side of this will be a Gutenberg block, just like we have core Search block? And this is the "frontend" side?

If so, see [`extensions/`](https://github.com/Automattic/jetpack/tree/808a541e94598c444bc89250e80e6c1cf565444f/extensions#readme) for how this code could be structured, built and deployed to .com as well.

Commit:
Add preact to dependencies, prune excess deps

Commit:
Merge search webpack config into main config

simison commented on _inc/search/components/api.js:
```diff
@@ -0,0 +1,14 @@
+	}
+
+	fetch( query ) {
+		return window.fetch( this.getAPIUrl( query ) );
```
See convo about using `apiFetch`/`fetch` and what you need to account for when bringing this code to .com:

https://github.com/Automattic/jetpack/pull/12823#pullrequestreview-256217084

import { addQueryArgs } from '@wordpress/url';

jsnmoon replied:
Let me try summarizing what the conversation linked above says:

1. We can't use `@wordpress/api-fetch` because it always sets a `X-WP-Nonce` header which WordPress.com disallows. 
2. If we use [`fetch`](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API), we avoid the `X-WP-Nonce` header issue.

Is this correct?

simison commented on _inc/search/components/search-results.jsx:
```diff
@@ -0,0 +1,27 @@
+		const { results = [], query } = this.props;
+		return (
+			<div>
+				<p>You are searching for "{ query }".</p>
```
Remember to include `__()` from `@wordpress/i18n` here.

simison commented on _inc/search/components/search-widget.jsx:
```diff
@@ -0,0 +1,78 @@
+						ref={ this.bindInput }
+					/>
+				</p>
+				<Portal into="#results">
```
You'll prolly need to change this ID attribute to something more unique.

ockham replied:
Why do we need a `Portal` here? (Can you maybe provide a screenshot to illustrate it?) Typically Portals should be reserved to control spatially disjoint pieces of the UI -- like a Gutenberg block controlling the elements in the editor sidebar (but we have higher-level interfaces for cases like that.)

I'd like us to be super-cautious about introducing `Portal`s -- since they allow us to break out of React's typical component hierarchy, they come with a risk of :spaghetti: UI code.

notnownikki replied:
Potential use case here is the search widget is in the sidebar or header, but the results appear in the main content area.

ockham replied:
That makes sense :thinking: Curious about some UI and implementation aspects though -- would the user add the widget to the sidebar, and a results block to a post's content? Or to a theme/page template? The [ID problem](https://github.com/Automattic/jetpack/pull/12741#discussion_r299356870) @simison has pointed out possibly highlights some conceptual issues with this. (E.g. I guess we can restrict the results block to appear only once in a post, but there could still be multiple posts with that block, visible at the same time in an archive-type page.) Maybe that's something complex enough to defer to a future iteration?

gibrown replied:
p3QzjZ-VL has mockups which hopefully clarifies some of this. I'm unsure if this needs an additional summary added to #13027 or not, but am pretty sure all of this code will change as we get more into the implementation.

simison commented on _inc/search/components/search-widget.jsx:
```diff
@@ -0,0 +1,78 @@
+				this.requestId++;
+				const requestId = this.requestId;
+				api
+					.fetch( query )
```
Remember to add error handling here.

simison commented on _inc/search/components/api.js:
```diff
@@ -0,0 +1,14 @@
+	}
+
+	fetch( query ) {
+		return window.fetch( this.getAPIUrl( query ) );
```
Remember that `window.fetch` isn't available in some browsers, notably IE11. You'll need to polyfill it.

jsnmoon replied:
In that vein, I'd recommend ponyfilling with a library like [unfetch](https://www.npmjs.com/package/unfetch)  which consumes only [554 bytes](https://bundlephobia.com/result?p=unfetch@4.1.0) gzipped.

simison commented on _inc/search/components/api.js:
```diff
@@ -0,0 +1,14 @@
+		let url = 'https://public-api.wordpress.com/rest/v1.3/sites/20115252/search?q=';
+		url += encodeURIComponent( query );
+		url += '&fields=author';
+		return url;
```
You could also use `addQueryArgs` from `@wordpress/url` but I understand if you wouldn't want to add dependencies to frontend. :-)

simison commented on _inc/search/components/search-widget.jsx:
```diff
@@ -0,0 +1,78 @@
+			<div>
+				<p>
+					<input
+						type="text"
```
Would type "search" work here better?

jsnmoon replied:
👍

simison commented on _inc/search/src/index.jsx:
```diff
@@ -0,0 +1,30 @@
+	render(
+		<SearchWidget
+			initialValue={ initialValue }
+			grabFocus={ grabFocus }
```
What should happen when there are multiple widgets requesting to `grabFocus`?

notnownikki replied:
Whichever fires off last would grab the focus. But, that's up to the higher level application to manage. `grabFocus` is there so if we're using progressive enhancement, we can maintain the focus on the search field if the user already has the search field focused. If the page is saying there are multiple input fields with focus at the same time, then the logic probably needs fixing there :)

simison commented on _inc/search/components/search-widget.jsx:
```diff
@@ -0,0 +1,78 @@
+						return response.json();
+					} )
+					.then( json => {
+						if ( this.requestId === requestId ) {
```
Can you elaborate on why this mechanism? Wouldn't `debounce` protect only latest request running? Or if not, should you implement cancellation so that you won't have multiple requests on fly simultaneously?

notnownikki replied:
I've seen folks think that `debounce` protects against this, and it doesn't. Cancellation of previous fetches would probably be easier to follow though.

Slow connections and unreliable connections are the reason you can't rely on `debounce` to protect against race conditions. If request 1 fires off, and then half a second later (a typical debounce timeout) request 2 fires off, there's no guarantee they'll resolve in that order. On fast networks they _probably_ would, but on mobile and other networks that are quite variable in speed, there's no guarantee.

ockham commented on _inc/search/components/search-widget.jsx:
```diff
@@ -0,0 +1,78 @@
+		} else {
+			this.setState( { results: [] } );
+		}
+	}
```
I'd strongly recommend replacing `JetpackSearchAPI` and the request handling in this component by a `@wordpress/data`-style store, which can take care of most of these things for us (by virtue of the concept of resolvers in particular).

See the example at https://github.com/WordPress/gutenberg/blob/c5f3c445ae858ba38a23d3f2981bcc8258565a9a/packages/data/README.md#registering-a-store -- it's similar enough to what we're trying to do here that you can essentially copy-paste it with a few modifications. LMK if you want me to help with that.

(Another -- real-live -- example is here: https://github.com/Automattic/jetpack/pull/12823/files#diff-1e7fffc443acdbb76875dedea0598f3a)

notnownikki replied:
Yes resolvers take care of a lot of problems, but the tradeoff is bundle size. As this is a front end widget that is potentially loaded for every visitor, we wanted to keep the dependencies as light as possible. I'll take a look, but as I remember the data package pulls in quite a lot of stuff.

ockham replied:
Thanks! Bummer about the data package pulling in lots of stuff :slightly_frowning_face: TBH, I'm always worried that a workaround pattern like this will proliferate through copy-paste to places where it isn't strictly needed (e.g. editor view), when we have an established pattern.

It's really unfortunate that we find ourselves working around the tooling that Core (the `@wordpress/` packages, specifically) provides us with, out of concern for bundle size :slightly_frowning_face: This seems very unwieldy; I'd personally rather use that tooling, and try to address its size upstream.

notnownikki replied:
Yeah, unfortunately the base react bundle is 113k (uncompressed) which is before we bring in any of the `@wordpress` packages. That's quite a hit for a widget that's potentially going on every page for every user.

(also, this is exactly the type of discussion I wanted to figure out here, I know the widget itself needs work and enhancements ;) but the actual inclusion mechanism is my worry)

simison commented on _inc/search/components/search-widget.jsx:
```diff
@@ -0,0 +1,78 @@
+						ref={ this.bindInput }
+					/>
+				</p>
+				<Portal into="#results">
```
What would happen with multiple search widgets on one page? Should you include a running unique integer?

notnownikki replied:
These ids should be props passed to the widget, definitely.

gibrown replied:
Added to #13027

ockham commented on package.json:
```diff
@@ -129,6 +129,8 @@
 		"photon": "2.0.1",
 		"plugin-error": "1.0.1",
+		"preact": "8.4.2",
+		"preact-portal": "1.1.3",
```
We should be really deliberate about using `preact` instead of `React` here. I'm not sure what it'd mean for JSX transpilation, and if it's worth the kilobytes saved. There's also quite some risk in interoperability with `@wordpress/element`. This isn't a decision we should take lightly; is there a P2 thread where we've discussed this and come to a conclusion? (I think it's conceivable that we end up using React on the frontend at some point, as more and more Gutenberg blocks require some level of interactivity on the frontend.)

notnownikki replied:
Yes, here https://jetpackcrew.wordpress.com/2019/05/16/react-for-front-end-code/

The end result was using preact produced a bundle an order of magnitude smaller that react, and as this is a front end widget potentially loaded for every visitor, it was worth it.

Personally, I'd like to use react because we'd be able to reusable the `@wordpress/*` packages, but the conclusion was bundle size mattered more.

ockham replied:
Ah right, I vaguely recall that discussion, thanks for referencing :sweat_smile: 

Yeah, sigh. I wonder how much longer we'll limbo around using React on the frontend if we even introduce `preact` as a drop-in. :slightly_frowning_face:

jeherve reviewed (dismissed):
Do you think you could add some testing instructions to the PR? You can rely on the PR template to create a PR description that will include as many details as possible:
https://raw.githubusercontent.com/Automattic/jetpack/master/.github/PULL_REQUEST_TEMPLATE.md

Commit:
Enqueue Jetpack Search prototype script given flag

Commit:
Add Search build output to gitignore

Commit:
Only import the debounce package from lodash

Commit:
Add 30KB performance budget for Search prototype

Commit:
Use search input instead of text input

Commit:
Use unfetch ponyfill

Commit:
Add translation

Commit:
Apply performance budget only for prod builds

Commit:
Move child component imports into parent

gibrown commented:
@jeherve @simison @ockham thanks for all the reviews and feedback. We've moved some of the issues off into #13027 because lots of things are left to be implemented and a lot of the JS code will need to get reworked from this initial code. We've also now added the code that loads the js assets on the front end and have some testing to make sure the bundle is not too large.

I think we are all onboard with the general direction of using preact and the set of libraries we've chosen, which I think is the most important piece of this PR.

Could you all take another look and see if we are missing anything else for getting us a stable framework for moving forward with additional development? If there is additional JS app coding ideas/feedback, then let's add them to #13027 in the interests of not losing them, but also not holding things up at this stage.

The travis-ci test failure above I don't really understand. I tried also running phpunit on the master branch and am also seeing some failures (though different ones). None of these really seem related to this PR, so I'm hoping we can ignore them or you have some thoughts.

Thanks!

jeherve commented on .gitignore:
```diff
@@ -47,6 +47,7 @@ phpcs.xml
 ## Things we will need in release branches
 /_inc/blocks
 /_inc/build
+/_inc/search/dist/
```
That's an helpful addition, thank you. 

We would also need to add the files we don't want to ship in a release in `.svnignore`:
https://github.com/Automattic/jetpack/blob/ee0fcec521f94cb2596ea4a8879350ca098457ae/.svnignore#L52

Could you speak a bit about your choice of a new subdirectory of `_inc/search` for the built files? I wonder if it may be easier to decide what to ship if the built files were in `_inc/build` instead?

gibrown replied:
The entire `_inc/search` directory is new. I only added `dist` because that is how the existing code was structured. I don't think there is any reason not to move them into `_inc/build` especially if that helps us hook into the existing Jetpack JS CDN in any way (do you know how that works?).

Is moving them your recommendation? I was also concerned that the build dir may be mostly for the dashboard client build.

simison replied:
In longer-term it could be helpful to get everything under `_inc/build` (including blocks), using one Webpack config file just with several different entries.

> I was also concerned that the build dir may be mostly for the dashboard client build.

I think that is mostly true but only because the admin dash has been the only bit of ES6 code in Jetpack until recently. It can be more clear to think the separation of code in terms of ES6 vs ES5 instead of "admin dash vs other features". 

If it's easy to move files to `_inc/build`, I'd go ahead and do that and not set too strong precedence of having different sub-directories for each feature.

Blocks are now built with the separate config to a separate folder because we couldn't untangle how building admin dash works and combine the two pipelines into one (https://github.com/Automattic/jetpack/issues/11745). Some work has since happened to make it easier tho.

Block extensions also have a mechanism to build those files for .com on Phabricator commits, so piggybacking on that pipeline could be useful to you.

jeherve replied:
> I don't think there is any reason not to move them into _inc/build especially if that helps us hook into the existing Jetpack JS CDN in any way (do you know how that works?).

You done have to worry about Jetpack's file CDN; as long as the files are in SVN, they'll be available and loaded via the CDN when the module is on, regardless of their location in the plugin structure.

jeherve replied:
> Is moving them your recommendation?

Yeah, I think it would be okay to move `_inc/search/dist` to `_inc/build`, alongside the other built files, where it will just be shipped with the other files with no additional changes needed.

gibrown replied:
Great, thanks for the feedback. I've added moving the built files to `_inc/build` to that open issue, so we'll get around to it.

jeherve commented on modules/search/class.jetpack-search.php:
```diff
@@ -186,13 +186,40 @@ public function init_hooks() {
+	/**
+	 * Loads assets for Jetpack Search Prototype featuring Search As You Type experience.
+	 */
+	public function load_assets() {
```
Will we want to load this on all pages by default, or only when a searchform is present?

gibrown replied:
When activated then needs to be on all pages.

jeherve reviewed (approved):
I rebased so I could test all this. This looks good to me, and works as advertized so far.

That's not blocking for now but I think we would really benefit from having more comments everywhere about what this all does, how it works, why we chose Preact. I think it's important since there will be non-a12s looking at this as well, trying to understand it all. It may also help us when we come back to it later. 

I think a readme file in the new `_inc/search` directory would be the minimum. You can follow the lead of that readme for example:
https://github.com/Automattic/jetpack/blob/9576f913b9aaa32cbcec38bb1ad9e7dabdb97372/_inc/client/README.md

jeherve commented:
Leaving the Needs Review label so @ockham and @simison can take another look since they looked at it earlier.
